pmlc/dialect/tile/transforms/pad_ranges.cc

```diff
@@ -0,0 +1,148 @@
+// Copyright 2019, Intel Corporation
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "llvm/Support/MathExtras.h"
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/padding.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/util/ident.h"
+#include "pmlc/util/logging.h"
+#include "pmlc/util/strides.h"
+
+namespace pmlc::dialect::tile {
+
+using namespace mlir; // NOLINT
+
+namespace {
+
+struct PadRangesPass : public PadRangesBase<PadRangesPass> {
+  int64_t maxPowerOfTwo = 16;
+  int64_t minPowerOfTwo = 8;
+  int64_t maxIncreasePercent = 5;
+  void runOnFunction() final;
+};
+
+void PadRangesPass::runOnFunction() {
+  auto func = getFunction();
+  func.walk([&](ContractionOp op) {
+    // Skip some cases where the padding pass can't operate.
+    if (op.getNumSymbols()) {
+      op.emitRemark("padding cannot be run on symbolic contractions");
+      return;
+    }
+    if (!op.lowerBounds() || !op.upperBounds()) {
+      op.emitRemark("contraction bounds must be computed");
+      return;
+    }
+    // Extract which dimensions of the contraction are which output dimension
+    SmallVector<unsigned, 4> outIdx;
+    unsigned outRank = op.getResultType().getRank();
+    for (unsigned i = 0; i < outRank; i++) {
+      auto outDim = op.sink().getResult(i).dyn_cast<AffineDimExpr>();
+      if (!outDim) {
+        op.emitRemark("Invalid output access");
+        return;
+      }
+      outIdx[i] = outDim.getPosition();
+    }
+    // Verify bounds are simple
+    unsigned indexCount = op.lowerBounds()->getNumResults();
+    SmallVector<int64_t, 6> ranges;
+    for (unsigned i = 0; i < indexCount; i++) {
+      auto lowerConst =
+          op.lowerBounds()->getResult(i).dyn_cast<AffineConstantExpr>();
+      auto upperConst =
+          op.upperBounds()->getResult(i).dyn_cast<AffineConstantExpr>();
+      if (!lowerConst || lowerConst.getValue() != 0) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      if (!upperConst) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      ranges.push_back(upperConst.getValue() + 1);
```

Comment:
Why not push upper - lower + 1 so we can handle the lower != 0 case?  Given `ranges` is signed we can handle this case?  Do we only ever expect 0 for a lower bound?

Reply:
Some of my comments got hidden by GitHub and therefore not seen by @jbruestle.  Commenting here in hopes that they will be made visible.

Reply:
Handling non-zero lower bound requires applying the proper offset and some additional analysis.  However, the prior passes/lowering should have already normalized lower bound to zero, so it will never fail in the existing pipeline.  I still check to be sure, since technically the IR allows non-zero lower bound.